File: src/com/sierrawireless/avphone/task/SyncWithAvTask.java
package com.sierrawireless.avphone.task;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import net.airvantage.model.AirVantageException;
import net.airvantage.model.Application;
import net.airvantage.model.AvError;
import net.airvantage.model.AvSystem;
import net.airvantage.model.UserRights;
import android.app.Activity;
import android.os.AsyncTask;
import android.text.Html;
import android.util.Log;

import com.sierrawireless.avphone.DeviceInfo;
import com.sierrawireless.avphone.MainActivity;
import com.sierrawireless.avphone.R;
import com.sierrawireless.avphone.message.IMessageDisplayer;
import com.sierrawireless.avphone.model.AvPhoneApplication;
import com.sierrawireless.avphone.model.CustomDataLabels;

public class SyncWithAvTask extends AsyncTask<SyncWithAvParams, SyncProgress, AvError> {

    private IApplicationClient applicationClient;

    private ISystemClient systemClient;

    private IAlertRuleClient alertRuleClient;

    private List<SyncWithAvListener> syncListeners = new ArrayList<SyncWithAvListener>();

    private IUserClient userClient;

    public SyncWithAvTask(IApplicationClient applicationClient, ISystemClient systemClient,
            IAlertRuleClient alertRuleClient, IUserClient userClient) {
        this.applicationClient = applicationClient;
        this.systemClient = systemClient;
        this.alertRuleClient = alertRuleClient;
        this.userClient = userClient;
    }

    public void addProgressListener(SyncWithAvListener listener) {
        this.syncListeners.add(listener);
    }

    @Override
    protected AvError doInBackground(SyncWithAvParams... params) {

        try {

            publishProgress(SyncProgress.CHECKING_RIGHTS);

            List<String> missingRights = userClient.checkRights();

            if (!missingRights.isEmpty()) {
                return new AvError(AvError.MISSING_RIGHTS, missingRights);
            }

            SyncWithAvParams syncParams = params[0];

            String serialNumber = syncParams.deviceId;
            String imei = syncParams.imei;
            String mqttPassword = syncParams.mqttPassword;
            CustomDataLabels customData = syncParams.customData;

            publishProgress(SyncProgress.CHECKING_APPLICATION);

            Application application = this.applicationClient.ensureApplicationExists(serialNumber);

            publishProgress(SyncProgress.CHECKING_SYSTEM);

            net.airvantage.model.AvSystem system = this.systemClient.getSystem(serialNumber);
            if (system == null) {

                publishProgress(SyncProgress.CREATING_SYSTEM);

                system = systemClient.createSystem(serialNumber, imei, mqttPassword, application.uid);
            }

            publishProgress(SyncProgress.CHECKING_ALERT_RULE);

            net.airvantage.model.AlertRule alertRule = this.alertRuleClient.getAlertRule(serialNumber);
            if (alertRule == null) {

                publishProgress(SyncProgress.CREATING_ALERT_RULE);

                this.alertRuleClient.createAlertRule(serialNumber, system.uid, application.uid);
            } else {

                publishProgress(SyncProgress.UPDATING_ALERT_RULE);

                // The alert rule exists, but was linked to another system
                // This probably means that the system was deleted, we'll update the alert rule.
                alertRule = this.alertRuleClient.updateAlertRule(alertRule.uid, serialNumber, system.uid,
                        application.uid);

            }

            publishProgress(SyncProgress.UPDATING_APPLICATION);

            this.applicationClient.setApplicationData(application.uid, customData);

            if (!hasApplication(system, application)) {

                publishProgress(SyncProgress.ADDING_APPLICATION);

                this.applicationClient.addApplication(system, application);
            }

            publishProgress(SyncProgress.DONE);

            return null;
        } catch (AirVantageException e) {
            publishProgress(SyncProgress.DONE);
            return e.getError();
        } catch (IOException e) {
            Log.e(MainActivity.class.getName(), "Error when trying to synchronize with server", e);
            publishProgress(SyncProgress.DONE);
            return new AvError("unkown.error");
        }

    }

    @Override
    protected void onPostExecute(AvError result) {
        super.onPostExecute(result);
        for (SyncWithAvListener listener : syncListeners) {
            listener.onSynced(result);
        }
    }

    private boolean hasApplication(AvSystem system, Application application) {
        boolean found = false;
        if (system.applications != null) {
            for (Application app : system.applications) {
                if (app.uid.equals(application.uid)) {
                    found = true;
                }
            }
        }
        return found;
    }

    public void showResult(AvError error, IMessageDisplayer displayer, Activity context) {

        if (error != null) {
            if (error.missingRights()) {
                String message = missingRightsMessage(error, context);
                displayer.showErrorMessage(Html.fromHtml(message));
            } else if (error.systemAlreadyExists()) {
                displayer.showError(R.string.sync_error_system_exists);
            } else if (error.gatewayAlreadyExists()) {
                displayer.showError(R.string.sync_error_gateway_exists);
            } else if (error.applicationAlreadyUsed()) {
                displayer.showError(R.string.sync_error_app_exists,
                        AvPhoneApplication.appType(DeviceInfo.getUniqueId(context)));
            } else if (error.tooManyAlerRules()) {
                displayer.showError(R.string.sync_error_too_many_rules);
            } else if (error.cantCreateApplication()) {
                displayer.showError(R.string.sync_error_no_right_create_application);
            } else if (error.cantCreateSystem()) {
                displayer.showError(R.string.sync_error_no_right_create_system);
            } else if (error.cantCreateAlertRule()) {
                displayer.showError(R.string.sync_error_no_right_create_alert_rule);
            } else if (error.cantUpdateApplication()) {
                displayer.showError(R.string.sync_error_no_right_update_app);
            } else if (error.cantUpdateSystem()) {
                displayer.showError(R.string.sync_error_no_right_update_system);
            } else if (error.forbidden()) {
                String method = error.errorParameters.get(0);
                String url = error.errorParameters.get(1);
                displayer.showError(R.string.sync_error_forbidden, method, url);
            } else {
                displayer.showError(R.string.sync_error_unexpected, error.error);
            }
        } else {
            displayer.showSuccess(R.string.sync_success);
        }
    }

    private String missingRightsMessage(AvError error, Activity context) {
        List<String> missingRights = error.errorParameters;
        StringBuilder message = new StringBuilder();
        message.append(context.getText(R.string.auth_not_enough_rights));
        message.append("<br/>");
        message.append(context.getText(R.string.auth_missing_rights));
        message.append("<br/>");

        for (String missingRight : missingRights) {
            message.append("&#8226; ");
            message.append(UserRights.asString(missingRight, context));
            message.append("<br/>");
        }
        return message.toString();
    }

}

File: mainActivity/src/main/java/com/sierrawireless/avphone/task/SyncWithAvTask.kt
package com.sierrawireless.avphone.task

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.util.Log
import com.crashlytics.android.Crashlytics
import com.sierrawireless.avphone.tools.DeviceInfo
import com.sierrawireless.avphone.activity.MainActivity
import com.sierrawireless.avphone.ObjectsManager
import com.sierrawireless.avphone.R
import com.sierrawireless.avphone.message.IMessageDisplayer
import net.airvantage.model.AirVantageException
import net.airvantage.model.Application
import net.airvantage.model.AvError
import net.airvantage.model.AvSystem
import java.io.IOException
import java.util.*


typealias SyncWithAvListener = (SyncWithAvResult) -> Unit

open class SyncWithAvTask internal constructor(private val applicationClient: IApplicationClient, private val systemClient: ISystemClient,
                                               private val alertRuleClient: IAlertRuleClient, private val userClient: IUserClient, @field:SuppressLint("StaticFieldLeak")
                                               protected val context: Context) : AvPhoneTask<SyncWithAvParams, SyncProgress, SyncWithAvResult>() {

    private val syncListeners = ArrayList<SyncWithAvListener>()
    var deviceName:String? = null

    fun addProgressListener(listener: SyncWithAvListener) {
        this.syncListeners.add(listener)
    }

    @SuppressLint("DefaultLocale")
    override fun doInBackground(vararg params: SyncWithAvParams): SyncWithAvResult {

        try {

            publishProgress(SyncProgress.CHECKING_RIGHTS)

            val missingRights = userClient.checkRights()
            if (!missingRights.isEmpty()) {
                return SyncWithAvResult(AvError(AvError.MISSING_RIGHTS, missingRights))
            }

            val systemType: String?
            val syncParams = params[0]
            val user = userClient.user
            val imei = syncParams.imei
            val iccid = syncParams.iccid
            deviceName = syncParams.deviceName
            val mqttPassword = syncParams.mqttPassword
            val objectsManager = ObjectsManager.getInstance()

            systemType = objectsManager.savedObjectName

            // For emulator and iOs compatibility sake, using generated serial.
            val serialNumber = DeviceInfo.generateSerial(user!!.uid!!)

            // Save Device serial in context
            if (context is MainActivity) {
                context.systemSerial = serialNumber
            }

            publishProgress(SyncProgress.CHECKING_APPLICATION)

            val application = this.applicationClient.ensureApplicationExists(deviceName!!)

            publishProgress(SyncProgress.CHECKING_SYSTEM)

            var system: net.airvantage.model.AvSystem? = this.systemClient.getSystem(serialNumber, systemType!!)
            if (system == null) {

                publishProgress(SyncProgress.CREATING_SYSTEM)

                system = systemClient.createSystem(serialNumber, iccid!!, systemType, mqttPassword!!, application.uid!!, deviceName!!, user.name!!, imei!!)
            }

            publishProgress(SyncProgress.CHECKING_ALERT_RULE)

            val alertRule = this.alertRuleClient.getAlertRule(serialNumber, system!!)
            if (alertRule == null) {

                publishProgress(SyncProgress.CREATING_ALERT_RULE)

                this.alertRuleClient.createAlertRule(application.uid!!, system!!)
            }

            publishProgress(SyncProgress.UPDATING_APPLICATION)

            this.applicationClient.setApplicationData(application.uid!!, objectsManager.savecObject.datas, objectsManager.savecObject.name!!)

            if (!hasApplication(system, application)) {

                publishProgress(SyncProgress.ADDING_APPLICATION)

                this.applicationClient.addApplication(system, application)
            }

            publishProgress(SyncProgress.DONE)

            return SyncWithAvResult(system, user)

        } catch (e: AirVantageException) {
            publishProgress(SyncProgress.DONE)
            return SyncWithAvResult(e.error!!)
        } catch (e: IOException) {
            Crashlytics.logException(e)
            Log.e(MainActivity::class.java.name, "Error when trying to synchronize with server", e)
            publishProgress(SyncProgress.DONE)
            return SyncWithAvResult(AvError("unkown.error"))
        }

    }

    override fun onPostExecute(result: SyncWithAvResult) {
        super.onPostExecute(result)
        for (listener in syncListeners) {
            listener.invoke(result)
        }
    }

    private fun hasApplication(system: AvSystem, application: Application): Boolean {
        var found = false
        if (system.applications != null) {
            system.applications!!
                    .filter { it.uid == application.uid }
                    .forEach { found = true }
        }
        return found
    }

    fun showResult(result: SyncWithAvResult, displayer: IMessageDisplayer, context: Activity) {

        if (result.isError) {
            val error = result.error
            displayTaskError(error!!, displayer, context, userClient, deviceName!!)
        } else {
            displayer.showSuccess(R.string.sync_success)
        }
    }

}
